torch/csrc/distributed/rpc/tensorpipe_agent.cpp

```diff
@@ -0,0 +1,367 @@
+#include <caffe2/torch/csrc/distributed/rpc/tensorpipe_agent.h>
+
+#include <torch/csrc/distributed/rpc/request_callback_impl.h>
+#include <torch/csrc/distributed/rpc/utils.h>
+
+#include <tensorpipe/channel/basic/basic.h>
+#include <tensorpipe/channel/cma/cma.h>
+#include <tensorpipe/transport/shm/context.h>
+#include <tensorpipe/transport/uv/context.h>
+
+namespace torch {
+namespace distributed {
+namespace rpc {
+
+using clock = std::chrono::high_resolution_clock;
+using nss = std::chrono::nanoseconds;
+
+std::atomic<uint32_t> TensorPipeAgent::nextShmAddressID_(0);
+
+TensorPipeAgent::TensorPipeAgent(
+    worker_id_t selfId,
+    std::string selfName,
+    std::shared_ptr<::c10d::Store> addressStore,
+    TensorPipeRpcBackendOptions opts)
+    : RpcAgent(
+          WorkerInfo(selfName, selfId),
+          std::make_unique<RequestCallbackImpl>(),
+          opts.rpcTimeout),
+      context_(tensorpipe::Context::create()),
+      addressStore_(std::move(addressStore)),
+      opts_(std::move(opts)) {
+  // Generate the maps for once.
+  for (const std::pair<std::string, worker_id_t>& kv : opts_.workerNameToId) {
```

Comment:
@mrshenli In the `init_rpc` function (see [docs here](https://pytorch.org/docs/stable/rpc.html#torch.distributed.rpc.init_rpc)) the rank seems to be optional (since it takes -1 as default), although all examples do actually pass in a rank. So can RPC agents assume that a valid rank will always be provided, or do they need to be able to handle the case of each node just knowing its name (and the world size)?

Reply:
Yes, we can expect all callers to pass in a valid rank to `init_rpc` for now. But it is unclear to me whether we can expect the same thing when support dynamic worker join/leaves after enable elasticity. We probably can let the store to assign rank so sth then.